k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/FolderSelectedStateCommand.java

```diff
@@ -0,0 +1,207 @@
+        boolean condstoreSupported = connection.isCondstoreCapable();
+        if (condstoreSupported) {
+            return LENGTH_LIMIT_WITH_CONDSTORE;
+        } else {
+            return LENGTH_LIMIT_WITHOUT_CONDSTORE;
+        }
+    }
+
+    static abstract class Builder<C extends FolderSelectedStateCommand, B extends Builder<C, B>> {
+        C command;
+        B builder;
+
+        abstract C createCommand();
+        abstract B createBuilder();
+
+        public Builder() {
+            command = createCommand();
+            builder = createBuilder();
+        }
+
+        public B idSet(Collection<Long> idSet) {
+            if (idSet != null) {
+                command.idSet = new TreeSet<>(idSet);
+            } else {
+                command.idSet = null;
+            }
+            return builder;
+        }
+
+        B addId(Long id) {
+            if (command.idSet == null) {
+                command.idSet = new TreeSet<>();
+            }
+            command.idSet.add(id);
+            return builder;
+        }
+
+        B idRanges(List<ContiguousIdGroup> idGroups) {
+            command.idGroups = idGroups;
+            return builder;
+        }
+
+        public B addIdGroup(Long start, Long end) {
+            if (command.idGroups == null) {
+                command.idGroups = new ArrayList<>();
+            }
+            command.idGroups.add(new ContiguousIdGroup(start, end));
+            return builder;
+        }
+
+        public B allIds(boolean allIds) {
+            if (allIds) {
+                command.idSet = null;
+                command.idGroups = Collections.singletonList(new ContiguousIdGroup(ContiguousIdGroup.FIRST_ID,
+                        ContiguousIdGroup.LAST_ID));
+            }
+            return builder;
+        }
+
+        public B onlyHighestId(boolean onlyHighestId) {
+            if (onlyHighestId) {
+                command.idSet = null;
+                command.idGroups = Collections.singletonList(new ContiguousIdGroup(ContiguousIdGroup.LAST_ID,
+                        ContiguousIdGroup.LAST_ID));
+            }
+            return builder;
+        }
+
+        public C build() {
+            return command;
+        }
+    }
+
+    static class ContiguousIdGroup {
+        static final long FIRST_ID = 1L;
+        static final long LAST_ID = Long.MAX_VALUE;
+
+        private Long start;
+        private Long end;
+
+        ContiguousIdGroup(Long start, Long end) {
+            if (start <= end) {
+                this.start = start;
+                this.end = end;
+            } else {
+                this.start = end;
+                this.end = start;
+            }
+        }
+
+        public Long getStart() {
+            return start;
+        }
+
+        public Long getEnd() {
+            return end;
+        }
+
+        @Override
+        public String toString() {
+            if (start == LAST_ID && end == LAST_ID) {
```

Comment:
Nope, this is meant to cover the method where UID SEARCH is used to find the highest UID. The case 1:* is covered by the block below this.